### PR TITLE
FLINK-16024 [jdbc-connector] JDBC Filter Pushdown

### PR DESCRIPTION
**Pull request by qingwei91**
## Note
This PR applies the same change as #20140 #20136, and might be useful to be reviewed by the same person.

## What is the purpose of the change

Implement Filter Pushdown in JDBC Connector Source, this is an optimization that will avoid scanning the whole table into Flink when we only need a subset of the table.

## Brief change log

* Implement a SQL Visitor that converts Flink SQL AST to JDBC SQL String
* Implement FilterPushdown in JDBCDynamicSource using the Visitor
* Add tests for both Visitor and JDBCDynamicSource

## Verifying this change

This change added tests and can be verified as follows:

* Added Integrated Test in JdbcDynamicTableSourceITCase to make sure there is no regression on SQL Filtering
* Added JdbcFilterPushdownVisitorTest to verify we are generating SQL correctly

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/config/PulsarClientFactory.java [hunk 1]
```diff
@@ -153,8 +153,8 @@ public static PulsarClient createClient(PulsarConfiguration configuration) {
     }
 
     /**
-     * PulsarAdmin shares almost the same configuration with PulsarClient, we use this
-     * creating method for creating PulsarAdmin instead of using {@link ClientConfigurationData}.
+     * PulsarAdmin shares almost the same configuration with PulsarClient, we use this creating
+     * method for creating PulsarAdmin instead of using {@link ClientConfigurationData}.
```
Comment (qingwei91) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/config/PulsarClientFactory.java:157]: This is unrelated to my change, but the CI complains about this

Comment (MartijnVisser) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/config/PulsarClientFactory.java:157]: This is being worked on via https://github.com/apache/flink/pull/20154

Comment (MartijnVisser) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/config/PulsarClientFactory.java:157]: @qingwei91 If you rebase, the issue will be resolved

**Comment by MartijnVisser** [comment]
@qingwei91 You're opening PRs against `release-1.15` but they should go towards `master` since these changes will need to end up in Flink 1.16.

**Comment by qingwei91** [comment]
Hi @MartijnVisser 
I've opened PR against master #20140 and against 1.14 #20136 

Will they be accepted? Or are those 2 versions in code freeze now?

I wish to update all three versions because we are using 1.14 internally in our organization, and we need the feature.

**Comment by MartijnVisser** [comment]
> I wish to update all three versions because we are using 1.14 internally in our organization, and we need the feature.

Unfortunately Flink doesn't add new features to already released versions, we can only release them in new minor versions.

**Comment by qingwei91** [comment]
@MartijnVisser oh, I see, I guess patch is only reserved for vulnerabilities or bug fixing?

That's unfortunate, I will just contribute to master for 1.16 then, thanks for explaining this.

**Pull request closed without merging**